### ISSUE TITLE
Pypy performs poorly on higher-order functions.

### ISSUE THREAD

**Issue by AgarwalPragy** [pypy/pypy#5146]
In the following example one would expect the 2 benchmarks to take the same amount of time

```py
import timeit

def f1(a, b): return a + b
def f2(a, b): return a + b

def hot_loop(func):
    total = 0
    for _ in range(1_000_000_000): total += func(1, 2)
    return total

print("f1:", timeit.timeit(lambda: hot_loop(f1), number=1))  # fast
print("f2:", timeit.timeit(lambda: hot_loop(f2), number=1))  # 5x slower
```

However, f2 is almost 5x slower than f1

It seems that Pypy jits the `hot_loop` function expecting the value of `func` to be equal to `f1`. But when `hot_loop(f2)` is called, the assumptions that Pypy made are violated, hence the code runs much slower.

This can be verified by adding a new function `hot_loop_2`

```py

import timeit

def f1(a, b): return a + b
def f2(a, b): return a + b

def hot_loop(func):
    total = 0
    for _ in range(1_000_000_000): total += func(1, 2)
    return total

def hot_loop_2(f):
    total = 0
    for _ in range(1_000_000_000): total += f(1, 2)
    return total

print("f1:", timeit.timeit(lambda: hot_loop(f1), number=1))                    # fast
print("f1 again:", timeit.timeit(lambda: hot_loop(f1), number=1))              # still fast
print("f2 with hot_loop:", timeit.timeit(lambda: hot_loop(f2), number=1))      # slow
print("f2 with hot_loop_2:", timeit.timeit(lambda: hot_loop_2(f2), number=1))  # fast!
```

## Why is this surprising?

I understand that if the type of a value changes inside a hot loop, it will be difficult for pypy to optimize the code.

However in the above examples the value (and its type) remains constant during the hot loop, and only changes b/w the hot loops.

Any higher order function would exhibit similar behavior, and thus will perform poorly with pypy

**Comment by cfbolz** [comment]
yeah, the jit specializes on the first function it sees, and the path for the second called function is less optimal. This is particularly pronounced for cases like your micro-benchmark, where both the higher order function and the inner function that you call are really small.

note however, that pypy is still easily 10x faster than cpython, even in the slow case.

do you have a real-world use cases where this matters or is it idle curiosity? it's fixable in theory, but it would be a significant chunk of work and it's not completely clear that it is worth it (particularly given that there are heuristics needed to identify this kind of situation, which might slow down other things).

**Comment by AgarwalPragy** [comment]
> note however, that pypy is still easily 10x faster than cpython, even in the slow case.

Agreed.

> do you have a real-world use cases where this matters or is it idle curiosity?

I don't have a specific use-case as such. Just that my code deals with a lot of higher oder functions, and I'm concerned if this will cause subtle performance issues.

I was benchmarking how different formulations of my higher order function performed with Pypy and ran into this.

> it's fixable in theory, but it would be a significant chunk of work

That's fair :)

Are there any guidelines that can help prevent similar jitting issues when working with higher order functions?

**Comment by cfbolz** [comment]
> my code deals with a lot of higher oder functions, and I'm concerned if this will cause subtle performance issues.

if you find a performance issue with your higher order functions "in the real world", that would be incredibly useful! Please report that, so we can decide based on actual use-cases how to optimize.
